src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs

```diff
@@ -228,41 +228,57 @@ public override bool DirectoryExists(string fullPath)
 
         public override IEnumerable<string> EnumeratePaths(string fullPath, string searchPattern, SearchOption searchOption, SearchTarget searchTarget)
         {
-            IEnumerable<IStorageItem> storageFiles = SynchronousResultOf(EnumerateFileQuery(fullPath, searchPattern, searchOption, searchTarget));
-
-            // Yield the paths of the StorageItems
-            foreach (IStorageItem item in storageFiles)
-            {
-                yield return item.Path;
-            }
+            IReadOnlyList<IStorageItem> storageFiles = SynchronousResultOf(EnumerateFileQuery(fullPath, searchPattern, searchOption, searchTarget));
+            return TranslatePaths(storageFiles);
         }
 
         public override IEnumerable<FileSystemInfo> EnumerateFileSystemInfos(string fullPath, string searchPattern, SearchOption searchOption, SearchTarget searchTarget)
         {
-            IEnumerable<IStorageItem> storageFiles = SynchronousResultOf(EnumerateFileQuery(fullPath, searchPattern, searchOption, searchTarget));
+            IReadOnlyList<IStorageItem> storageFiles = SynchronousResultOf(EnumerateFileQuery(fullPath, searchPattern, searchOption, searchTarget));
+            return TranslateFileSystemInfos(storageFiles);
+        }
 
+        /// <summary>
+        /// Translates IStorageItems into FileSystemInfos
+        /// </summary>
+        private static IEnumerable<FileSystemInfo> TranslateFileSystemInfos(IReadOnlyList<IStorageItem> storageFiles)
+        {
             // Translate the StorageFile results into FileSystemInfos
-            foreach (IStorageItem item in storageFiles)
+            int count = storageFiles.Count;
+            for (int i = 0; i < count; i++)
             {
-                if (item.IsOfType(StorageItemTypes.Folder))
+                if (storageFiles[i].IsOfType(StorageItemTypes.Folder))
```

Comment:
could keep a local copy to avoid calling the indexer twice. Don't know that it's worthwhile though.